cuBERT_topic_modelling/embedding_extraction.py

```diff
@@ -0,0 +1,120 @@
+from cudf.core.subword_tokenizer import SubwordTokenizer
+import torch
+from transformers import AutoModel
+from torch.utils.data import TensorDataset, DataLoader
+import transformers
+
+# Vocabulary is included in the root directory of this repo
+# however, below is the command to modify / update it -->
+# from cudf.utils.hash_vocab_utils import hash_vocab
+# hash_vocab('vocab.txt', 'voc_hash.txt')
+
+device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+
+# Mean Pooling - Take attention mask into account for correct averaging
+def mean_pooling(model_output, attention_mask):
+    """Function to implement mean pooling on top of the AutoModel
+    See: https://www.sbert.net/examples/applications/computing-embeddings/README.html#sentence-embeddings-with-transformers
+
+    Args:
+        model_output \
+            (transformers.BaseModelOutputWithPoolingAndCrossAttentions): BERT model
+        attention_mask (torch.Tensor): torch.Tensor representing attention
+        mask values
+
+    Returns:
+        [torch.Tensor]: correct averaging of attention mask
+    """
+    # First element of model_output contains all token embeddings
+    token_embeddings = model_output[0]
+    input_mask_expanded = attention_mask.\
+        unsqueeze(-1).\
+        expand(token_embeddings.size()).\
+        float()
+    sum_embeddings = torch.sum(token_embeddings * input_mask_expanded, 1)
+    sum_mask = torch.clamp(input_mask_expanded.sum(1), min=1e-9)
+    return sum_embeddings / sum_mask
+
+def create_embeddings(sentences):
+    """Creates the sentence embeddings using SentenceTransformer
+
+    Args:
+        sentences (cudf.Series[str]): a cuDF Series of Input strings
+
+    Returns:
+        embeddings (cupy.ndarray): corresponding sentence
+        embeddings for the strings passed
+    """
+
+    cudf_tokenizer = SubwordTokenizer(
+        'vocab/voc_hash.txt',
+        do_lower_case=True
+    )
+
+    # Tokenize sentences
+    encoded_input_cudf = cudf_tokenizer(
+        sentences,
+        max_length=128,
+        max_num_rows=len(sentences),
+        padding='max_length',
+        return_tensors='pt',
+        truncation=True
+    )
+
+    # Load AutoModel from huggingface model repository
+    model_gpu = AutoModel.from_pretrained(
+        "sentence-transformers/all-MiniLM-L6-v2"
+    ).to(device)
+
+    # Delete the key and associated values we do not need
+    del encoded_input_cudf['metadata']
+
+    batch_size = 64
+
+    dataset = TensorDataset(
+        encoded_input_cudf['input_ids'],
+        encoded_input_cudf['attention_mask']
+    )
+
+    dataloader = DataLoader(
+        dataset,
+        batch_size=batch_size
+    )
+
+    model_o_ls = []
+    model_pooler_output = []
+    with torch.no_grad():
+        for data in dataloader:
+            model_obj = model_gpu(**{
```

Comment:
@VibhuJawa I'm running into CUDA OOM issues while running my code on a dataset of size > 40K documents. I have triaged this issue to this loop. Things I have tried so far:

1. adding `del` statements to remove unused variables
2. using `torch.cuda.empty_cache()` - it doesn't go OOM but is extremely slow and not recommended to be used (https://discuss.pytorch.org/t/about-torch-cuda-empty-cache/34232/4)
3. I found a few places where they recommended to reduce the batch size, but that also didn't work out for me

This issue is blocking benchmarking, let me know if you have any thoughts.

Reply:
I am attaching a MRE here as well: https://gist.github.com/mayankanand007/62103291804f1b64ee700a29b3289aa2

Reply:
The problem here is that the size of  `model_obj.last_hidden_state` `torch.Size([64, 128, 384]` which is about `128` times greater than what we actually need which is embeddings `torch.Size([64, 384])` .   Essentially we cant save that much in memory.
We should do something like below where we gather the output of each batch directly. We just need to check correctness but  based on my interpretation of https://www.sbert.net/examples/applications/computing-embeddings/README.html#sentence-embeddings-with-transformers we should be fine. 

```python

def mean_pooling(model_output, attention_mask):
    """Function to implement mean pooling on top of the AutoModel
    See: https://www.sbert.net/examples/applications/computing-embeddings/README.html#sentence-embeddings-with-transformers
    Args:
        model_output \
            (transformers.BaseModelOutputWithPoolingAndCrossAttentions): BERT model
        attention_mask (torch.Tensor): torch.Tensor representing attention
        mask values
    Returns:
        [torch.Tensor]: correct averaging of attention mask
    """
    # First element of model_output contains all token embeddings
    token_embeddings = model_output[0]
    input_mask_expanded = attention_mask.\
        unsqueeze(-1).\
        expand(token_embeddings.size()).\
        float()
    sum_embeddings = torch.sum(token_embeddings * input_mask_expanded, 1)
    sum_mask = torch.clamp(input_mask_expanded.sum(1), min=1e-9)
    return sum_embeddings / sum_mask

input_ids =    torch.rand(size=[100000, 128]).int().to(device)
attention_mask = torch.rand(size=[100000, 128]).int().to(device)

dataset =  TensorDataset(input_ids, attention_mask)
dataloader = DataLoader(
    dataset,
    batch_size=batch_size
)
del input_ids,attention_mask



pooling_output_ls = []
with torch.no_grad():
    for b_input_ids,b_attention_mask in dataloader:
        model_obj = model_gpu(**{'input_ids':b_input_ids, 'attention_mask':b_attention_mask})
        pooling_output_ls.append(mean_pooling(model_obj,b_attention_mask))
        
pooling_output = torch.cat(pooling_output_ls)
```